beacon-chain/rpc/beacon_chain_server.go

```diff
@@ -238,24 +238,43 @@ func (bs *BeaconChainServer) GetChainHead(ctx context.Context, _ *ptypes.Empty)
 	}, nil
 }
 
-// ListValidatorBalances retrieves the validator balances for a given set of public key at
-// a specific epoch in time.
-//
-// TODO(#3064): Implement balances for a specific epoch. Current implementation returns latest balances,
-// this is blocked by DB refactor.
+// ListValidatorBalances retrieves the validator balances for a given set of public keys.
+// An optional Epoch parameter is provided to request historical validator balances from
+// archived, persistent data.
 func (bs *BeaconChainServer) ListValidatorBalances(
 	ctx context.Context,
 	req *ethpb.GetValidatorBalancesRequest) (*ethpb.ValidatorBalances, error) {
 
 	res := make([]*ethpb.ValidatorBalances_Balance, 0, len(req.PublicKeys)+len(req.Indices))
 	filtered := map[uint64]bool{} // track filtered validators to prevent duplication in the response.
 
-	headState, err := bs.beaconDB.HeadState(ctx)
-	if err != nil {
-		return nil, status.Errorf(codes.Internal, "could not retrieve head state: %v", err)
+	var requestingGenesis bool
+	var epoch uint64
+	switch q := req.QueryFilter.(type) {
+	case *ethpb.GetValidatorBalancesRequest_Epoch:
+		epoch = q.Epoch
+	case *ethpb.GetValidatorBalancesRequest_Genesis:
+		requestingGenesis = q.Genesis
+	default:
 	}
-	balances := headState.Balances
+
+	var balances []uint64
+	var err error
+	headState := bs.headFetcher.HeadState()
 	validators := headState.Validators
+	if requestingGenesis {
+		balances, err = bs.beaconDB.ArchivedBalances(ctx, 0 /* genesis epoch */)
+		if err != nil {
+			return nil, status.Errorf(codes.InvalidArgument, "could not retrieve balances for epoch %d", epoch)
```

Comment:
Don't use `epoch` for this error message. Could become messy to debug if someone specifies genesis and also gives an epoch number

Reply:
It's impossible to pass in genesis and an epoch number together based on the proto :)